### PR TITLE
WebGPURenderer: Initial Antialiasing

### PR DESCRIPTION
**Pull request by takahirox**
This PR adds initial antialiasing support to `WebGPURenderer`.

**API**

`WebGPURenderer` constructor takes `antialias` parameter. Default is `false`.

`const new renderer = WebGPURenderer( { antialias: true } );`

**Screenshot**

Left with antialiasing and right without antialiasing.

image

**Changes**

In `WebGL`, `HTMLCanvasElement.getContext()` takes `antialias` parameter and if it's `true` hardware antialiasing is enabled.

In `WebGPU`, `HTMLCanvasElement.getContext()` doesn't take `antialias` and we seem to need to manually setup MSAA (at least now).

### CODE REVIEW

File: examples/jsm/renderers/webgpu/WebGPURenderer.js [hunk 1]
```diff
@@ -114,13 +118,38 @@ class WebGPURenderer {
 
 			const renderTargetProperties = this._properties.get( renderTarget );
 
-			colorAttachment.attachment = renderTargetProperties.colorTextureGPU.createView();
-			depthStencilAttachment.attachment = renderTargetProperties.depthTextureGPU.createView();
+			if ( this._antialias ) {
+
+				colorAttachment.attachment = renderTargetProperties.colorBuffer.createView();
+				colorAttachment.resolveTarget = renderTargetProperties.colorTextureGPU.createView();
+				colorAttachment.storeOp = GPUStoreOp.Store;
```
Comment (Kangz) [inline examples/jsm/renderers/webgpu/WebGPURenderer.js:125]: This should be GPUStoreOp.Clear, unless you're planning on reusing the content of the multisampled texture. It helps a TON with memory bandwidth on tiler GPUs. The resolve target pixels will be stored, but multisampled pixels  will be forgotten and the texture marked for lazy-clearing internally in Dawn/Chromium.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js [hunk 2]
```diff
@@ -114,13 +118,38 @@ class WebGPURenderer {
...
+			depthStencilAttachment.attachment = renderTargetProperties.depthTextureGPU.createView();
 
 		} else {
 
-			colorAttachment.attachment = this._swapChain.getCurrentTexture().createView();
+			if ( this._antialias ) {
+
+				colorAttachment.attachment = this._colorBuffer.createView();
+				colorAttachment.resolveTarget = this._swapChain.getCurrentTexture().createView();
+				colorAttachment.storeOp = GPUStoreOp.Store;
```
Comment (Kangz) [inline examples/jsm/renderers/webgpu/WebGPURenderer.js:143]: Ditto

**Comment by Kangz** [comment]
WebGPU canvas textures will always have a sample count of 1 because it makes the expensive MSAA resolve step explicit compared to WebGL where you could accidentally trigger resolves in some cases.

File: examples/jsm/renderers/webgpu/WebGPUTextures.js [hunk 3]
```diff
@@ -184,6 +185,23 @@ class WebGPUTextures {
...
+					},
+					sampleCount: this.sampleCount,
+					format: GPUTextureFormat.BRGA8Unorm,
+					usage: GPUTextureUsage.OUTPUT_ATTACHMENT
+				});
+
+				renderTargetProperties.colorBuffer = colorBuffer;
+
+			}
+
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPUTextures.js:204]: I'm not sure about this bit. The render target already has a configured color buffer. 

I suggest we look at multisampled render targets in another PR and thus revert the changes to `WebGPUTextures` for now.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js [hunk 4]
```diff
@@ -71,6 +72,9 @@ class WebGPURenderer {
 
 		this._renderTarget = null;
 
+		this._antialias = !! parameters.antialias;
+		this._sampleCount = this._antialias ? 4 : 1; // Should be configurable?
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPURenderer.js:76]: I vote to make this configurable with a new parameter `sampleCount`.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js [hunk 5]
```diff
@@ -71,6 +72,9 @@ class WebGPURenderer {
 
 		this._renderTarget = null;
 
+		this._antialias = !! parameters.antialias;
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPURenderer.js:75]: In order to keep the amount of "private" properties small, I suggest to remove `_antialias` and `sampleCount`. Both informations are already available over the `parameters` property.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js (hunk 4 above)
Comment (takahirox) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js:76]: That sounds good to me. Which do you prefer?

1. renderer constructor takes `antialias` and `sampleCount`. `sampleCount` is effective only `antialias` is `true`. Default `sampleCount` is 4. Pros is maybe `antialias` option would be intuitive especially to existing users.

2. renderer constructor takes only `sampleCount`. That it is greater than 1 means antialiasing is enabled. Pros is less the number of option parameters would be good.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js (hunk 5 above)
Comment (takahirox) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js:75]: That sounds good to me. I overlooked renderer keeps `parameters`.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js (hunk 4 above)
Comment (Mugen87) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js:76]: I vote for option 1.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js (hunk 1 above)
Comment (takahirox) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js]: Thanks for the comment. Updated.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js (hunk 4 above)
Comment (takahirox) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js]: Thanks, updated.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js (hunk 5 above)
Comment (takahirox) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js]: Thanks, updated.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js [hunk 6]
```diff
@@ -23,7 +23,7 @@ class WebGPURenderer {
 		// public
 
 		this.domElement = ( parameters.canvas !== undefined ) ? parameters.canvas : document.createElementNS( 'http://www.w3.org/1999/xhtml', 'canvas' );
-		this.parameters = parameters;
+		this.parameters = Object.assign( {}, parameters );
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPURenderer.js:26]: Do we need this code although the class already uses a default parameter? Meaning:
```js
constructor( parameters = {} ) {
```

Comment (takahirox) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js:26]: [I update some fields of `parameters` to ensure](https://github.com/mrdoob/three.js/pull/20281/files#diff-b2e471696b2cc2a9842ff64fe2ec36bbR166). I want to avoid having an effect to object user passed. So it is for copying object. But no strong opinion.

File: examples/jsm/renderers/webgpu/WebGPUTextures.js (hunk 3 above)
Comment (takahirox) [reply examples/jsm/renderers/webgpu/WebGPUTextures.js]: [WebGPURenderPipelines creates render pipeline with parameters.sampleCount with this change.](https://github.com/mrdoob/three.js/pull/20281/files#diff-b2e471696b2cc2a9842ff64fe2ec36bbR166) Attachment seems to need to have the same number sampleCount but I want `renderTarget.texture` to be non-multisampled texture because `WebGPURenderer` doesn't support multisampled-texture as input yet. So `.textureGPU` for resolveTarget (output, non-multisampled) and `.colorBuffer` for attachment (multisampled).

But I'm not sure if this is the best approach and I want to revisit later. So agreed with reverting for now. I added console error so far for render target + antialiasing.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js (hunk 6 above)
Comment (Mugen87) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js:26]: Ah, I see now the reason for this! Make sense 👍 .

**Review by Mugen87 — APPROVED** [review]

File: examples/jsm/renderers/webgpu/WebGPURenderer.js [hunk 7]
```diff
@@ -71,6 +72,20 @@ class WebGPURenderer {
 
 		this._renderTarget = null;
 
+		// ensure some parameters
+
+		this.parameters.antialias = !! this.parameters.antialias;
```
Comment (mrdoob) [inline examples/jsm/renderers/webgpu/WebGPURenderer.js:77]: I think this is easier to read:

```js
this.parameters.antialias = this.parameters.antialias === true;
```

Comment (takahirox) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js]: Updated, thanks

**Comment by mrdoob** [comment]
Thanks!

File: examples/jsm/renderers/webgpu/WebGPURenderer.js (hunk 1 above)
Comment (Mugen87) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js]: @takahirox I had to remove the modifications to `storeOp` since enabling antialiasing produced a red screen. Would be great if you can revisit this setting: 8e70942d186d761d892a9307a39faddb1c28172f

Comment (takahirox) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js]: Can you share a demo or code snippet which causes the problem? Here I enabled antialias in webgpu sandbox example but couldn't reproduce it.

Comment (Mugen87) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js]: Strange, it does not work on my system. I just add the code back and set `antialias` to `true` in the example.

Comment (takahirox) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js]: Hm it could be platform or hardware specific issue? Working fine on my Windows10 + Canary(87.0.4257.0) + intel(R) HD graphics 530.

Anyways, using default for now is ok to me because it still works on my system, too. Let's revisit later if needed.

Comment (Mugen87) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js]: Okay, I'll try to check the demo tomorrow on a Windows 10 notebook.
